### PR TITLE
Feature: make retry strategy exponential backoff multiplier configurable

### PR DESCRIPTION
**Pull request by Velfi**
_NOTE: I plan to allow for user-defined retry strategies in a separate PR._

## Motivation and Context

aws-sdk-rust#567

## Description
Updated SDK Client retry behavior to allow for a configurable backoff multiplier.

Previously, the backoff multiplier (named `r` in the code) was set to 2 seconds. This is not an ideal default for services like DynamoDB that expect clients to quickly retry failed request attempts. Now, users can set quicker (or slower) multipliers according to their needs.

```rust
#[tokio::main]
async fn main() -> Result<(), aws_sdk_dynamodb::Error> {
    let retry_config = aws_smithy_types::retry::RetryConfigBuilder::new()
        .max_attempts(4)
        .backoff_multiplier(Duration::from_millis(20));

    let shared_config = aws_config::from_env()
        .retry_config(retry_config)
        .load()
        .await;

    let client = aws_sdk_dynamodb::Client::new(&shared_config);

    // Given the 20ms backoff multiplier, and assuming this request fails 3 times before succeeding,
    // the first retry would take place between 0-20ms after the initial request,
    // the second retry would take place between 0-40ms after the first retry,
    // and the third retry would take place between 0-80ms after the second retry.
    let request = client
        .put_item()
        .table_name("users")
        .item("username", "Velfi")
        .item("account_type", "Developer")
        .send().await?;

    Ok(())
}
```

**For a request that gets retried 3 times, when base is 1 and initial_backoff is 2 seconds:**
- the first retry will occur after 0 to 2 seconds
- the second retry will occur after 0 to 4 seconds
- the third retry will occur after 0 to 8 seconds

**For a request that gets retried 3 times, when base is 1 and initial_backoff is 30 milliseconds:**
- the first retry will occur after 0 to 30 milliseconds
- the second retry will occur after 0 to 60 milliseconds
- the third retry will occur after 0 to 120 milliseconds

## Testing

I wrote some test for the backoff calculation and ran existing retry tests

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/aws-smithy-client/src/retry.rs [hunk 1]
```diff
@@ -57,6 +58,7 @@ pub struct Config {
     no_retry_increment: usize,
     timeout_retry_cost: usize,
     max_attempts: u32,
+    backoff_multiplier: Duration,
```
Comment (rcoh) [inline rust-runtime/aws-smithy-client/src/retry.rs:61]: so my first reaction is that a duration being a multiple is a little confusing. Is this more like the initial delay?

File: rust-runtime/aws-smithy-client/src/retry.rs [hunk 2]
```diff
@@ -532,4 +565,34 @@ mod test {
         assert!(no_retry.is_none());
         assert_eq!(policy.retry_quota(), 480);
     }
+
+    #[test]
+    fn calculate_exponential_backoff_n_is_one() {
+        let n = 1.0;
+
+        for (expected, iter) in [(n, 0u32), (2.0, 1), (4.0, 2)] {
```
Comment (rcoh) [inline rust-runtime/aws-smithy-client/src/retry.rs:573]: this was surprisingly hard to load into my brain—does `.enumerate()` work here?

File: rust-runtime/aws-smithy-client/src/retry.rs [hunk 3]
```diff
@@ -532,4 +565,34 @@ mod test {
         assert!(no_retry.is_none());
         assert_eq!(policy.retry_quota(), 480);
     }
+
+    #[test]
+    fn calculate_exponential_backoff_n_is_one() {
+        let n = 1.0;
```
Comment (rcoh) [inline rust-runtime/aws-smithy-client/src/retry.rs:571]: can you add a comment explaining what `n` is here?

**Review by rcoh — COMMENTED** [review]
this looks great! the one change I'd make is considering renaming `with_backoff_multiplier` to `initial_backoff`—it's what we call it in the interior code and I think it's a bit more grokable. But totally open to discuss this

File: rust-runtime/aws-smithy-client/src/retry.rs (hunk 1 above)
Comment (Velfi) [reply rust-runtime/aws-smithy-client/src/retry.rs:61]: Your suggestion to rename this one is a good one. It is a thing that we multiply the base exponential backoff duration by but I agree calling it initial_backoff will aid users in constructing a more helpful mental model of how it works.

File: rust-runtime/aws-smithy-client/src/retry.rs (hunk 2 above)
Comment (Velfi) [reply rust-runtime/aws-smithy-client/src/retry.rs:573]: yeah it does, I'll switch it

File: CHANGELOG.next.toml [hunk 4]
```diff
@@ -36,3 +36,53 @@ message = "Re-export aws_types::SdkConfig in aws_config"
...
+(named `r` in the code) was set to 2 seconds. This is not an ideal default for services that expect clients to quickly
+retry failed request attempts. Now, users can set quicker (or slower) multipliers according to their needs.
+"""
+references = ["aws-sdk-rust#567"]
+meta = { "breaking" = false, "tada" = true, "bug" = false }
+author = "Velfi"
+
+[[aws-sdk-rust]]
+message = """
+Updated SDK Client retry behavior to allow for a configurable backoff multiplier. Previously, the backoff multiplier
```
Comment (rcoh) [inline CHANGELOG.next.toml:52]: backoff multiplier -> initial delay

**Review by rcoh — APPROVED** [review]
Nice! LGTM

**Pull request merged**